### DISCUSSION TITLE
How to configurate the Gulp plugins dynaminally (inside pipeline individually for each file)?

### DISCUSSION THREAD

**Discussion by TokugawaTakeshi** [gulpjs/gulp#2688] [Help] [answered]
I started this discussion after [has not been answered in Stack Overflow](https://stackoverflow.com/q/72392347/4818123) even during bounty. But I still don't believe that what I want is impossible.

Consider the the example of standard usage of [gulp-postcss](https://github.com/postcss/gulp-postcss) plugin:

```javascript
gulp.task('default', function () {
    var plugins = [nested];
    return gulp.src('in.sss')
        .pipe(postcss(plugins, { parser: sugarss }))
        .pipe(gulp.dest('out'));
});
```

Besides this statis cofiguration scenario, **gulp-postcss** allows to cofigurate itself dynamically:

```typescript
pipe(gulpPostCSS((file: Vinyl): { plugins: Array<unknown>; } => ({
  plugins: [
    Autoprefixer,
    CSS_Nano({
      preset: [
        "default",
        {
          // Vinyl allows to add custom properties like "someBooleanFlag"
          discardComments: file.someBooleanFlag
        }
      ]
    })
  ]
}))).
```

But the [gulp-pug](https://github.com/gulp-community/gulp-pug#readme) plugin, for example, does not provide the callback allows to make the dynamic configuration dependent on Vinyl file:

```typescript
declare function GulpPug(options?: GulpPug.Options): Transform;
```

How to config this plugin an other plugins which does not provide the callback depending on specific Vinyl file?

**Comment by phated** [comment, marked as answer]
It is generally expected that you will use [gulp-if](https://www.npmjs.com/package/gulp-if) to match vinyl files and provide different plugin streams. That "function" config object in gulp-postcss is really a recommended pattern.

**Reply by TokugawaTakeshi** [reply]
@phated 

Thank you for the answer!

**gulp-if** works perfectly when we need to use some plugin conditionally, but what about conditional configuration of the specific plugin?

In below case, I need to pass the variable **pageStatesSimulations** of object type to Pug template. 
It's value is being stored in the instance of Vinyl file because this variable is specific-page dependent.

```javascript
exports.views = () => {
  return src('./src/*.pug')
    .pipe(
      pug({
        locals: {
          pageStatesSimulations: {} /* get from "file.statesSimulations" */
        }
      })
    )
    .pipe(dest('./dist'));
};
```

There also could be other specific file dependent cofiguration, for example, **filters**.

Would you please to teach me how to access to Vinyl file in this case?

**Reply by phated** [reply]
That's what [gulp-data](https://www.npmjs.com/package/gulp-data) is for. Use that instead of trying to use static locals.

**Reply by TokugawaTakeshi** [reply]
@TokugawaTakeshi 

Thank you again.

You have recommended the appropriate solution for my case.
However, because the currect topics is about generic solution for the arbitrary plugin, I as sorry, but I can't finish this discussion now.

In below example, the `stringOptionExample1` are not data which could be passed via plugins like **gulp-data** - it is the option which must colud be set depending on some cumstom property of Vinyl file. Some generic solution for this, or "basically it is impossible"?

```javascript
module.exports = () => {
  return src('./src/*.arbitrary')
    .pipe(
      arbitraryPlugin({
        stringOptionExample1: /* get from file.stringOptionExample */ || "DEFAULT_VALUE",
        stringOptionExample2: /* get from file.booleanOptoinsExample */ ? "ALPHA": "BRAVO",
        numericOptionExample: /* get from file.numertionOptionExample */ || 6 ,
        objectTypeOption: /* get from file.numertionOptionExample */,
        functionOption: /* get from file.functionOption */
      })
    )
    .pipe(dest('./dist'));
};
```

**Reply by phated** [reply]
@TokugawaTakeshi Why does `arbitraryPlugin` need to get properties from `filestringOptionExample`? Instead, they could get data from `file.data`, such as `file.data.filestringOptionExample`. The `arbitraryPlugin` can implement gulp-data support.

Other plugins have solved this by injecting themselves into the pipeline twice, such as gulp-sourcemaps (before gulp supported sourcemaps by default), which required `sourcemap.init()` to add the `.sourceMap` property and then `sourcemap.write()` to read the `.sourceMap` property and create a vinyl file.

In vinyl-fs, we use the [resolve-options](https://www.npmjs.com/package/resolve-options) package to resolve an option against a static value or function that receives a vinyl file. We've not recommended this pattern to plugin authors, but maybe it could be used within the ecosystem too.
